## Codio Exercise 8.2: Using `PolynomialFeatures`

**Estimated time: 60 minutes**

**Total Points: 20 Points**


This activity focuses on using the scikitlearn transformer `PolynomialFeatures`.  As seen in video 8.4, you can use this transformer to create the modified DataFrame with appropriate column names using the `.get_feature_names_out()` method on the fit transformer.  You will focus on building second, third, and fourth degree polynomial models using `PolynomialFeatures`, and converting the results to pandas DataFrames.

## Index:

 - [Problem 1](#Problem-1)
 - [Problem 2](#Problem-2)
 - [Problem 3](#Problem-3)
 - [Problem 4](#Problem-4)

In [1]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import PolynomialFeatures
from sklearn.feature_extraction.text import CountVectorizer

### The Data

Again, the automobile dataset is used.  You will build the additional features using the `horsepower` column of the data.  

In [2]:
auto = pd.read_csv("data/auto.csv")

In [3]:
auto.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
0  18.0          8         307.0       130.0    3504          12.0    70   
1  15.0          8         350.0       165.0    3693          11.5    70   
2  18.0          8         318.0       150.0    3436          11.0    70   
3  16.0          8         304.0       150.0    3433          12.0    70   
4  17.0          8         302.0       140.0    3449          10.5    70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

[Back to top](#Index:) 

## Problem 1

### Creating Quadratic Features

**5 Points**

Using the `horsepower` column, instantiate, fit, and transform a `PolynomialFeatures` transformer.  Instantiate the transformer and use the `degree` argument to create second degree polynomial features.  Assign your transformer to the variable `pfeatures`, and the transformed features as a numpy array to `quad_features` below.  Note that the `PolynomialFeatures` transformer expects a 2-dimensional input; one way to accomplish this is to use two sets of brackets to select a single column -- `auto[[colname]]`.

In [4]:
### GRADED

pfeatures = PolynomialFeatures(degree=2)
quad_features = pfeatures.fit_transform(auto[["horsepower"]])

# Answer check
print(type(quad_features))

<class 'numpy.ndarray'>


[Back to top](#Index:) 

## Problem 2

### Creating the DataFrame

**5 Points**

Using the transformed array, create a DataFrame of the transformed data.  As in the lectures, use the `get_feature_names_out()` method of your fit transformer `pfeatures` from above.  Drop the bias term from the DataFrame so that you only have the columns `horsepower` and `horsepower^2`.  Assign your response as a DataFrame to the variable `poly_features_df` below.

In [5]:
### GRADED

poly_features_df = pd.DataFrame(
    quad_features,
    columns=pfeatures.get_feature_names_out(),
).drop(columns="1")

# Answer check
print(poly_features_df.shape)
poly_features_df.head()

(392, 2)


horsepower  horsepower^2
0       130.0       16900.0
1       165.0       27225.0
2       150.0       22500.0
3       150.0       22500.0
4       140.0       19600.0

[Back to top](#Index:) 

## Problem 3

### DataFrame with Cubic Features

**5 Points**

Now, use a transformer to create a DataFrame with features for a third degree or cubic polynomial model.  Do this by setting the `degree=3` in your transformer.  As before, drop the bias term so that your final DataFrame has the shape `(392, 3)` and has feature names `horsepower`, `horsepower^2`, and `horsepower^3`.  Assign your results as a DataFrame to `cubic_features_df` below.

In [6]:
### GRADED

pfeatures = PolynomialFeatures(degree=3)
cubic_features = pfeatures.fit_transform(auto[["horsepower"]])
cubic_features_df = pd.DataFrame(
    cubic_features,
    columns=pfeatures.get_feature_names_out(),
).drop(columns="1")

# Answer check
print(cubic_features_df.shape)
cubic_features_df.head()

(392, 3)


horsepower  horsepower^2  horsepower^3
0       130.0       16900.0     2197000.0
1       165.0       27225.0     4492125.0
2       150.0       22500.0     3375000.0
3       150.0       22500.0     3375000.0
4       140.0       19600.0     2744000.0

[Back to top](#Index:) 

## Problem 4

### Experimenting with Multiple Features

**5 Points**

Now, experiment with building polynomial features for multiple columns.  Specifically, use a transformer to create a DataFrame of quadratic features (`degree = 2`) for the columns `horsepower` and `weight`.  Drop the bias term as before and examine the column names.  Note the existence of a new column `horsepower weight`.  Assign your transformed data as a DataFrame to `two_feature_poly_df` below.

In [7]:
### GRADED

pfeatures = PolynomialFeatures(degree=2)
quad_two_features = pfeatures.fit_transform(auto[["horsepower", "weight"]])
two_feature_poly_df = pd.DataFrame(
    quad_two_features,
    columns=pfeatures.get_feature_names_out(),
).drop(columns="1")

# Answer check
print(two_feature_poly_df.shape)
two_feature_poly_df.head()

(392, 5)


horsepower  weight  horsepower^2  horsepower weight    weight^2
0       130.0  3504.0       16900.0           455520.0  12278016.0
1       165.0  3693.0       27225.0           609345.0  13638249.0
2       150.0  3436.0       22500.0           515400.0  11806096.0
3       150.0  3433.0       22500.0           514950.0  11785489.0
4       140.0  3449.0       19600.0           482860.0  11895601.0

#### Summary

Now that you have the hang of using `PolynomialFeatures`, you will combine the transformer with an estimator using scikitlearn's pipeline utilities.  As demonstrated in the videos, the pipeline is a handy abstraction for combining the data transformations with the model in a single object.  This is especially handy when making predictions with new data points.